### ISSUE TITLE
 When it will support  3D model?

### ISSUE THREAD

**Issue by nicolae1992** [mapbox/mapbox-gl-native#14494]
When it will support  3D GLTF or other formats in mapbox android sdk

**Android versions:** 
**Device models:** 
**Mapbox SDK versions:**

**Comment by tobrun** [comment]
This is something we are thinking about exposing an API around though atm can't give you a concrete eta. The formats are still TBD but ideally it would be an open format. We will follow up with more concrete issues in the near future, closing this issue as answered for now.

**Comment by noway** [comment]
@tobrun Is there anywhere where we can vote to prioritise that feature? We are using mapbox and will need 3d map annotations very soon.

**Comment by nicolae1992** [comment]
@noway Thx, This task is very important

**Comment by astojilj** [comment]
@noway @nicolae1992

mapbox-gl-js demo https://bl.ocks.org/ryanbaumann/b9725e79b5b1742d92ada00ddbf5c3d3 uses Three.js for loading GLTF model. We don't plan to include GLTF model loader to mapbox code - hopefully you were able to find a loader to use. Please notify if following similar approach (custom layer e.g. https://github.com/mapbox/mapbox-gl-js/pull/7977 and https://bl.ocks.org/ryanbaumann/b9725e79b5b1742d92ada00ddbf5c3d3) doesn't work for you.

**Comment by nicolae1992** [comment]
@astojilj this link is js, but needed for Android SDK

**Comment by astojilj** [comment]
> @astojilj this link is js, but needed for Android SDK

Yes, mapbox-gl-js demo is useful to explain the approach: similarly, you'd need to load GLTF model using loader of your choice and use custom layer approach, as mentioned in https://github.com/mapbox/mapbox-gl-native/issues/14494#issuecomment-525216947

**Comment by noway** [comment]
@astojilj I'm using mapbox for react native, so ideally I need the 3d model layer support in `react-native-mapbox-gl/maps` package. I assume it would be something along the lined of 
```javascript
<MapboxGL.MapView>
  <MapboxGL.ThreeDLayer shape={shape} meshes={meshes}>
  </MapboxGL.ThreeDLayer>
</MapboxGL.MapView>
```
or something similar. I don't think I have capacity to build that feature out for `react-native-mapbox-gl/maps` package by my own. I heard that you guys are not supporting this package, so it might be on @nitaliano to port this to RN world. A fully figured out and documented Android/iOS support though is a good first step. I'm not sure if your mapbox-gl-js example would work for `mapbox-sdk-ios`/`mapbox-sdk-android` as it is?

**Comment by luckkyzhou** [comment]
> > @astojilj this link is js, but needed for Android SDK
> 
> Yes, mapbox-gl-js demo is useful to explain the approach: similarly, you'd need to load GLTF model using loader of your choice and use custom layer approach, as mentioned in [#14494 (comment)](https://github.com/mapbox/mapbox-gl-native/issues/14494#issuecomment-525216947)

Hi, I have found a java loader to load GLTF model. But now I need the documentation of custom layer approach for android SDK. I couldn't find how to custom layer with android, so is there any documentation about this?

**Comment by NickLambert95** [comment]
@tobrun Has there been an update on 3D objects in android yet? We're trying to put 3D cars on our map and have successfuly achieved it through the web gl but not on android or iOS

**Comment by noway** [comment]
Can this be reopened?

**Comment by tobrun** [comment]
cc @chloekraw

**Comment by chloekraw** [comment]
Hi folks, thanks for all your interest in this feature. Support for 3D objects is something we are exploring; there's no timeline for its release. We'll keep you posted.

**Comment by NickLambert95** [comment]
> Hi folks, thanks for all your interest in this feature. Support for 3D objects is something we are exploring; there's no timeline for its release. We'll keep you posted.

Hey @chloekraw not to be rude but this is the exact same answer we've heard over a year ago. Is there any updates on this? There are a lot of benefits and uses this feature can bring and we really need this for our app. Are there any work arounds that may accomplish this? 

Thanks

**Comment by leedejun** [comment]
I have finished loading 3d model with mapbox gl native, implement a custom layer that loads the 3D model, referencing the ASSIMp library.  As shown in figure：

<img width="892" alt="building" src="https://user-images.githubusercontent.com/11495799/88636943-20d65c80-d0ec-11ea-9b41-72672c2ccb96.png">

**Comment by NickLambert95** [comment]
> @chloekraw @nicolae1992 @NickLambert95
> 
> I have finished loading 3d model with mapbox gl native, implement a custom layer that loads the 3D model, referencing the assimp library.
> 
> @NickLambert95 You could check out my comment:
> 
> [#13157 (comment)](https://github.com/mapbox/mapbox-gl-native/issues/13157#issuecomment-608253844)

Hey @leedejun how exactly did you do this? Can I see your code that you edited? Thanks

**Comment by janweigel** [comment]
Hi @leedejun,
Could you please share the code you wrote to achieve 3D object layers for iOS and Android?
We need the same feature and there's no update for any official support for more than a year. 
Any help appreciated. 
Thank you!

**Comment by NickLambert95** [comment]
@janweigel He never got back to me either but it appears he made this in React Native (which is the web version of mapbox)

**Comment by patrickkempff** [comment]
> @janweigel He never got back to me either but it appears he made this in React Native (which is the web version of mapbox)

React Native does not use the web version of mapbox afaik

**Comment by NickLambert95** [comment]
@patrickkempff  Maybe not react native then. The mapbox GL version (which works on web) allows you to add GLTF files as 3D models. There isn't a workaround I'm aware of for iOS and Android

**Comment by janweigel** [comment]
If I get the post of @leedejun correctly then he modified the c++ code to add that feature to the native android and ios version and not the gl js version of mapbox.
That's exactly what I'm looking for.

**Comment by NickLambert95** [comment]
@janweigel Yeah I desperately need that function too. Let me know if he ever gets back to you and I'll do the same for you

**Comment by leedejun** [comment]
@patrickkempff @NickLambert95 @janweigel  I understand your problem, I solved it，I achieved it by modifying the custom layer of mapbox gl native in Mac, linux, android, windows, and used the Assimp library.  But I can't share my code, I must share my code after I leave the current company.  I would like share my method with you.
**the method detail**:
1. The Assimp is able to read the model files; 
2. Implement mapbox custom layer to draw 3d model;
3. Add a function for Convert latitude and longitude to Opengl coordinates, Rang: [-1.0,1.0], .

**Comment by janweigel** [comment]
Hi @leedejun,
Thanks for sharing that info. 
I'll see if I can implement the native part for a custom layer myself.

**Comment by Amrutha1991** [comment]
Hi @janweigel @leedejun 
Thanks for the information.
It would be helpful if anyone share the snippet for, custom layer to draw 3d model using mapbox native code.

Thanks in advance

**Comment by fabdrol** [comment]
It'd be even more useful if the code was shared with Mapbox, so @chloekraw can implement this officially

**Comment by NickLambert95** [comment]
Anyone have an update on this topic? Any solutions? @janweigel @fabdrol @patrickkempff @tobrun @luckkyzhou @leedejun @chloekraw

**Comment by Cal-um** [comment]
Also requesting 3D model support on iOS and Android.

**Comment by janweigel** [comment]
@NickLambert95 and @Cal-um 
We switched to Unity using the Mapbox Unity SDK, as it wasn't possible to wait for those features.

**Comment by NickLambert95** [comment]
@janweigel Oh I see. Out of curiosity does Mapbox Unity come with the same navigation features? As in, could I build a full fledge app with the navigation SDK and directions API in it? I looked this up before but couldn't find any information as to whether Unity supported it

**Comment by Cal-um** [comment]
that's a good shout. Probably better for my use case as well. Just need to learn unity 😂

**Comment by ajbraudev** [comment]
We are in desperate need of this as well. Hoping to hear an update soon.

**Comment by JeyBi24** [comment]
YES PLEASE, ADD 3D OBJECTS support for ANDROID and IOS

**Comment by tobrun** [comment]
Sorry for the long delay on this, we have been hard at work and do appreciate you all reaching out and requesting this feature. Not sure if you noticed but we [announced the v10.0.0-beta](https://www.mapbox.com/blog/mapbox-mobile-sdk-v10-beta) last week. This release of the SDK, which has been rewritten from the ground up, adds support for [3D terrain](https://github.com/mapbox/mapbox-maps-android/blob/main/app/src/main/java/com/mapbox/maps/testapp/examples/terrain/Terrain3DShowcaseActivity.kt) and [3D model](https://github.com/mapbox/mapbox-maps-android/blob/main/app/src/main/java/com/mapbox/maps/testapp/examples/ModelLayerActivity.kt). 

The repositories for these libraries are:
 - https://github.com/mapbox/mapbox-maps-android
 - https://github.com/mapbox/mapbox-maps-ios

While it's still in beta, not production ready yet, we do invite you to try it out and provide feedback.

**Comment by astojilj** [comment]
iOS example using SceneKit: https://github.com/mapbox/mapbox-maps-ios/pull/154

**Comment by NickLambert95** [comment]
@tobrun you posted this comment, but I can no longer find it. I'm not seeing the support for 3D models on the links you sent

Sorry for the long delay on this, we have been hard at work and do appreciate you all reaching out and requesting this feature. Not sure if you noticed but we announced the v10.0.0-beta last week. This release of the SDK, which has been rewritten from the ground up, adds support for 3D terrain and 3D model.

The repositories for these libraries are:

https://github.com/mapbox/mapbox-maps-android
https://github.com/mapbox/mapbox-maps-ios
While it's still in beta, not production ready yet, we do invite you to try it out and provide feedback.

**Comment by ajbraudev** [comment]
Any update on this? We would be willing to contribute and collaborate on this issue if necessary. Would love to hear an update, even if it's bad news. Thanks Mapbox Team!

**Comment by ajbraudev** [comment]
@tobrun Any update on this? We would be willing to contribute and collaborate on this issue if necessary. Would love to hear an update, even if it's bad news. Thanks Mapbox Team!

**Comment by soheilnikbin** [comment]
Hi @astojilj , it seems that the Mapbox iOS SDK [3D Puck](https://github.com/mapbox/mapbox-maps-ios/blob/7ba6d499e565d65da9ad54b6f4348eecdbb505da/Apps/Examples/Examples/All%20Examples/Custom3DPuckExample.swift#L30) is using the private ModelLayer to render glTF objects. We are wondering why this functionality is not publicly available, as it would simplify our development. It seems there is a demand for this feature.

**Comment by b4rtt** [comment]
> Hi @astojilj , it seems that the Mapbox iOS SDK [3D Puck](https://github.com/mapbox/mapbox-maps-ios/blob/7ba6d499e565d65da9ad54b6f4348eecdbb505da/Apps/Examples/Examples/All%20Examples/Custom3DPuckExample.swift#L30) is using the private ModelLayer to render glTF objects. We are wondering why this functionality is not publicly available, as it would simplify our development. It seems there is a demand for this feature.

Yes. 👍🏻

**Comment by astojilj** [comment]
Hello,
We are working on this, as presented in https://www.mapbox.com/blog/mapbox-3d-navigation.

![image](https://user-images.githubusercontent.com/549216/216292680-5ce4f3bc-b33e-4dd2-8d24-441734286da2.png)

I cannot share implementation details or milestones now. Thanks.
